File: src/main/groovy/com/bugsnag/android/gradle/BugsnagReleasesTask.java
package com.bugsnag.android.gradle;

import com.android.build.gradle.api.BaseVariant;
import com.android.build.gradle.api.BaseVariantOutput;
import org.gradle.api.Action;
import org.gradle.api.DefaultTask;
import org.gradle.api.Project;
import org.gradle.api.logging.LogLevel;
import org.gradle.api.logging.Logger;
import org.gradle.api.tasks.TaskAction;
import org.gradle.process.ExecSpec;
import org.gradle.process.internal.ExecException;
import org.json.simple.JSONObject;
import org.xml.sax.SAXException;

import javax.xml.parsers.ParserConfigurationException;
import java.io.BufferedReader;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.Charset;
import java.util.Arrays;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

public class BugsnagReleasesTask extends DefaultTask {

    private static final Collection<String> VALID_VCS_PROVIDERS =
        Arrays.asList("github-enterprise", "bitbucket-server", "gitlab-onpremise", "bitbucket",
            "github", "gitlab");

    private static final String MK_OS_ARCH = "os.arch";
    private static final String MK_OS_NAME = "os.name";
    private static final String MK_OS_VERSION = "os.version";
    private static final String MK_JAVA_VERSION = "java.version";
    private static final String VCS_COMMAND = "git";
    private static final String CHARSET_UTF8 = "UTF-8";

    BaseVariantOutput variantOutput;
    BaseVariant variant;

    @TaskAction
    void fetchReleaseInfo() throws IOException, SAXException, ParserConfigurationException {
        Project project = getProject();
        final AndroidManifestInfo manifestInfo = BugsnagVariantOutputUtils.readManifestFile(project, variant, variantOutput);
        Logger logger = project.getLogger();

        if (!isValidPayload(manifestInfo.getApiKey(), manifestInfo.getVersionName())) {
            logger.warn("Must supply api key and version name for release task");
            return;
        }
        final BugsnagPluginExtension bugsnag = project.getExtensions().getByType(BugsnagPluginExtension.class);
        final JSONObject payload = generateJsonPayload(manifestInfo, bugsnag);
        String json = payload.toString();
        logger.debug("Releases Payload:\n" + json);

        new Call(project) {
            @Override
            public boolean makeApiCall() throws IOException {
                return deliverPayload(payload, manifestInfo, bugsnag);
            }
        }.execute();
    }

    private boolean deliverPayload(JSONObject payload, AndroidManifestInfo manifestInfo, BugsnagPluginExtension bugsnag) throws IOException {
        OutputStream os = null;
        Logger logger = getProject().getLogger();

        try {
            URL url = new URL(bugsnag.getReleasesEndpoint());
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();

            conn.setRequestMethod("POST");
            conn.setRequestProperty("Content-Type", "application/json");
            conn.setRequestProperty("Bugsnag-Api-Key", manifestInfo.getApiKey());
            conn.setReadTimeout(bugsnag.getRequestTimeoutMs());
            conn.setConnectTimeout(bugsnag.getRequestTimeoutMs());
            conn.setDoOutput(true);

            os = conn.getOutputStream();
            os.write(payload.toString().getBytes(CHARSET_UTF8));

            int statusCode = conn.getResponseCode();

            if (statusCode == 200) {
                logger.info("Uploaded release info to Bugsnag");
                return true;
            } else {
                BufferedReader reader = null;
                String line;

                try {
                    reader = new BufferedReader(new InputStreamReader(conn.getErrorStream()));
                    while ((line = reader.readLine()) != null) {
                        logger.error(line);
                    }
                    logger.warn("Release Request failed with statusCode " + statusCode);
                } finally {
                    if (reader != null) {
                        reader.close();
                    }
                }
                return false;
            }
        } catch (IOException e) {
            logger.error(bugsnag.getReleasesEndpoint());
            logger.error("Failed to POST request", e);
            return false;
        } finally {
            if (os != null) {
                os.close();
            }
        }
    }

    private JSONObject generateJsonPayload(AndroidManifestInfo manifestInfo, BugsnagPluginExtension bugsnag) {
        JSONObject root = new JSONObject();

        root.put("buildTool", "gradle-android");
        root.put("apiKey", manifestInfo.getApiKey());
        root.put("appVersion", manifestInfo.getVersionName());
        root.put("appVersionCode", manifestInfo.getVersionCode());

        String user;
        if (bugsnag.getBuilderName() != null) {
            user = bugsnag.getBuilderName();
        } else {
            user = runCmd("whoami");
        }
        root.put("builderName", user);

        root.put("metadata", generateMetadataJson(bugsnag));
        root.put("sourceControl", generateVcsJson(bugsnag));
        return root;
    }

    private JSONObject generateVcsJson(BugsnagPluginExtension bugsnag) {
        SourceControl sourceControl = AgpCompat.getSourceControl(getProject());
        String vcsUrl = null;
        String commitHash = null;
        String vcsProvider = null;

        if (sourceControl != null) {
            vcsUrl = sourceControl.getRepository();
            commitHash = sourceControl.getRevision();
            vcsProvider = sourceControl.getProvider();
        }

        if (vcsUrl == null) {
            vcsUrl = runCmd(VCS_COMMAND, "config", "--get", "remote.origin.url");
        }
        if (commitHash == null) {
            commitHash = runCmd(VCS_COMMAND, "rev-parse", "HEAD");
        }
        if (vcsProvider == null) {
            vcsProvider = parseProviderUrl(vcsUrl);
        }

        JSONObject sourceControlObj = new JSONObject();
        sourceControlObj.put("repository", vcsUrl);
        sourceControlObj.put("revision", commitHash);

        if (isValidVcsProvider(vcsProvider)) {
            sourceControlObj.put("provider", vcsProvider);
        }
        return sourceControlObj;
    }

    private JSONObject generateMetadataJson(BugsnagPluginExtension bugsnag) {
        Map<String, String> metadata = collectDefaultMetaData();
        Map<String, String> userData = bugsnag.getMetadata();

        if (userData != null) { // user values overwrite any defaults
            for (Map.Entry<String, String> entry : userData.entrySet()) {
                metadata.put(entry.getKey(), entry.getValue());
            }
        }

        JSONObject additionalInfo = new JSONObject();

        for (Map.Entry<String, String> entry : metadata.entrySet()) {
            additionalInfo.put(entry.getKey(), entry.getValue());
        }
        return additionalInfo;
    }

    private Map<String, String> collectDefaultMetaData() {
        String gradleVersion = getProject().getGradle().getGradleVersion();
        Map<String, String> metadata = new HashMap();
        metadata.put("os_arch", System.getProperty(MK_OS_ARCH));
        metadata.put("os_name", System.getProperty(MK_OS_NAME));
        metadata.put("os_version", System.getProperty(MK_OS_VERSION));
        metadata.put("java_version", System.getProperty(MK_JAVA_VERSION));
        metadata.put("gradle_version", gradleVersion);
        metadata.put("git_version", runCmd(VCS_COMMAND, "--version"));
        return metadata;
    }

    static boolean isValidPayload(String apiKey, String versionName) {
        return apiKey != null && versionName != null;
    }

    static boolean isValidVcsProvider(String provider) {
        return provider == null || VALID_VCS_PROVIDERS.contains(provider);
    }

    static String parseProviderUrl(String url) {
        if (url != null) {
            for (String provider : VALID_VCS_PROVIDERS) {
                if (url.contains(provider)) {
                    return provider;
                }
            }
        }
        return null;
    }

    /**
     * Runs a command on the shell
     * @param cmd the command (arguments must be separate strings)
     * @return the cmd output
     */
    private String runCmd(final String... cmd) {
        try {
            final ByteArrayOutputStream baos = new ByteArrayOutputStream();

            getProject().exec(new Action<ExecSpec>() {
                @Override
                public void execute(ExecSpec execSpec) {
                    execSpec.commandLine(cmd);
                    execSpec.setStandardOutput(baos);
                    getLogging().captureStandardError(LogLevel.INFO);
                }
            });
            return new String(baos.toByteArray(), Charset.forName(CHARSET_UTF8)).trim();
        } catch (ExecException ignored) {
            return null;
        }
    }
}

File: src/main/kotlin/com.bugsnag.android.gradle/BugsnagReleasesTask.kt
package com.bugsnag.android.gradle

import com.squareup.moshi.JsonClass
import okhttp3.OkHttpClient
import org.gradle.api.DefaultTask
import org.gradle.api.Project
import org.gradle.api.file.FileCollection
import org.gradle.api.file.RegularFileProperty
import org.gradle.api.logging.LogLevel
import org.gradle.api.model.ObjectFactory
import org.gradle.api.provider.MapProperty
import org.gradle.api.provider.Property
import org.gradle.api.provider.ProviderFactory
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.InputFile
import org.gradle.api.tasks.Internal
import org.gradle.api.tasks.InputFiles
import org.gradle.api.tasks.Optional
import org.gradle.api.tasks.OutputFile
import org.gradle.api.tasks.PathSensitive
import org.gradle.api.tasks.PathSensitivity.NONE
import org.gradle.api.tasks.TaskAction
import org.gradle.process.internal.ExecException
import org.gradle.util.VersionNumber
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory
import retrofit2.converter.scalars.ScalarsConverterFactory
import retrofit2.create
import retrofit2.http.Body
import retrofit2.http.Header
import retrofit2.http.POST
import retrofit2.http.Url
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.nio.charset.Charset
import java.time.Duration
import javax.inject.Inject

open class BugsnagReleasesTask @Inject constructor(
    objects: ObjectFactory,
    private val providerFactory: ProviderFactory
) : DefaultTask(), AndroidManifestInfoReceiver {

    init {
        group = BugsnagPlugin.GROUP_NAME
        description = "Assembles information about the build that will be sent to the releases API"
    }

    @get:Internal
    internal val uploadRequestClient: Property<UploadRequestClient> = objects.property(UploadRequestClient::class.java)

    @get:PathSensitive(NONE)
    @get:InputFile
    override val manifestInfoFile: RegularFileProperty = objects.fileProperty()

    @get:OutputFile
    val requestOutputFile: RegularFileProperty = objects.fileProperty()

    // should take the JVM + NDK mapping files as inputs because the manifestInfo will
    // not necessarily vary between different builds. it is not guaranteed that
    // either of these properties will be set so they are marked as optional.
    @get:PathSensitive(NONE)
    @get:InputFile
    @get:Optional
    val jvmMappingFileProperty: RegularFileProperty = objects.fileProperty()

    @get:PathSensitive(NONE)
    @get:InputFiles
    @get:Optional
    val ndkMappingFileProperty: Property<FileCollection> = objects.property(FileCollection::class.java)

    @get:Input
    val retryCount: Property<Int> = objects.property(Int::class.javaObjectType)

    @get:Input
    val timeoutMillis: Property<Long> = objects.property(Long::class.javaObjectType)

    @get:Input
    val releasesEndpoint: Property<String> = objects.property(String::class.java)

    @get:Optional
    @get:Input
    val builderName: Property<String> = objects.property(String::class.java)

    @get:Optional
    @get:Input
    val metadata: MapProperty<String, String> = objects.mapProperty(String::class.java, String::class.java)

    @get:Optional
    @get:Input
    val sourceControlProvider: Property<String> = objects.property(String::class.java)

    @get:Optional
    @get:Input
    val sourceControlRepository: Property<String> = objects.property(String::class.java)

    @get:Optional
    @get:Input
    val sourceControlRevision: Property<String> = objects.property(String::class.java)

    @get:Input
    @get:Optional
    val osArch: Property<String> = objects.property(String::class.java)

    @get:Input
    @get:Optional
    val osName: Property<String> = objects.property(String::class.java)

    @get:Input
    @get:Optional
    val osVersion: Property<String> = objects.property(String::class.java)

    @get:Input
    @get:Optional
    val javaVersion: Property<String> = objects.property(String::class.java)

    @get:Input
    @get:Optional
    val gradleVersion: Property<String> = objects.property(String::class.java)

    @get:Input
    @get:Optional
    val gitVersion: Property<String> = objects.property(String::class.java)

    @TaskAction
    fun fetchReleaseInfo() {
        val manifestInfo = parseManifestInfo()
        val payload = generateJsonPayload(manifestInfo)

        val response = uploadRequestClient.get().makeRequestIfNeeded(manifestInfo, payload.toString()) {
            logger.lifecycle("Bugsnag: Attempting upload to Releases API")
            lateinit var response: String
            object : Call(retryCount, logger) {
                override fun makeApiCall(): Boolean {
                    response = deliverPayload(payload, manifestInfo)
                    requestOutputFile.asFile.get().writeText(response)
                    logger.lifecycle("Bugsnag: Upload succeeded")
                    return true
                }
            }.execute()
            response
        }
        requestOutputFile.asFile.get().writeText(response)
        logger.lifecycle("Bugsnag: Releases request complete")
    }

    private fun deliverPayload(
        payload: ReleasePayload,
        manifestInfo: AndroidManifestInfo
    ): String {
        val timeoutDuration = Duration.ofMillis(timeoutMillis.get())
        val bugsnagService = createService(timeoutDuration)

        val response = try {
            bugsnagService.upload(
                releasesEndpoint.get(),
                apiKey = manifestInfo.apiKey,
                payload = payload
            ).execute()
        } catch (e: IOException) {
            throw IllegalStateException("Request to Bugsnag Releases API failed, aborting build.", e)
        }
        return readRequestResponse(response)
    }

    private fun readRequestResponse(response: Response<String>): String {
        val statusCode = response.code()
        val success = statusCode == 200
        val responseData = when {
            success -> response.body().orEmpty()
            else -> response.errorBody()?.string().orEmpty()
        }
        return when {
            success -> responseData
            else -> {
                logger.error(responseData)
                throw IllegalStateException("Request to Bugsnag Releases API failed, aborting build.")
            }
        }
    }

    private fun generateJsonPayload(manifestInfo: AndroidManifestInfo): ReleasePayload {
        return ReleasePayload(
            buildTool = "gradle-android",
            apiKey = manifestInfo.apiKey,
            appVersion = manifestInfo.versionName,
            appVersionCode = manifestInfo.versionCode,
            metadata = generateMetadataJson(),
            sourceControl = generateVcsJson(),
            builderName = if (builderName.isPresent) {
                builderName.get()
            } else {
                runCmd("whoami")
            }
        )
    }

    private fun generateVcsJson(): Map<String, String?> {
        var vcsUrl = sourceControlRepository.orNull
        var commitHash = sourceControlRevision.orNull
        var vcsProvider = sourceControlProvider.orNull
        if (vcsUrl == null) {
            vcsUrl = runCmd(VCS_COMMAND, "config", "--get", "remote.origin.url")
        }
        if (commitHash == null) {
            commitHash = runCmd(VCS_COMMAND, "rev-parse", "HEAD")
        }
        if (vcsProvider == null) {
            vcsProvider = parseProviderUrl(vcsUrl)
        }
        val sourceControlObj = mutableMapOf<String, String?>()
        sourceControlObj["repository"] = vcsUrl
        sourceControlObj["revision"] = commitHash
        if (isValidVcsProvider(vcsProvider)) {
            sourceControlObj["provider"] = vcsProvider
        }
        return sourceControlObj
    }

    private fun generateMetadataJson(): Map<String, String?> {
        val metadataMap = mutableMapOf<String, String?>()
        collectDefaultMetaData(metadataMap)
        metadataMap.putAll(metadata.orNull.orEmpty())
        return metadataMap.toMap()
    }

    private fun collectDefaultMetaData(map: MutableMap<String, String?>) {
        map["os_arch"] = osArch.orNull
        map["os_name"] = osName.orNull
        map["os_version"] = osVersion.orNull
        map["java_version"] = javaVersion.orNull
        map["gradle_version"] = gradleVersion.orNull
        map["git_version"] = gitVersion.orNull
    }

    /**
     * Runs a command on the shell
     * @param cmd the command (arguments must be separate strings)
     * @return the cmd output
     */
    private fun runCmd(vararg cmd: String): String? {
        return try {
            val baos = ByteArrayOutputStream()
            project.exec { execSpec ->
                execSpec.commandLine(*cmd)
                execSpec.standardOutput = baos
                logging.captureStandardError(LogLevel.INFO)
            }
            String(baos.toByteArray(), Charset.forName(CHARSET_UTF8)).trim { it <= ' ' }
        } catch (ignored: ExecException) {
            null
        }
    }

    internal fun configureMetadata() {
        val gradleVersionNumber = gradleVersion.orNull?.let {
            gradleVersion.set(it)
            VersionNumber.parse(it)
        }
        gitVersion.set(providerFactory.provider { runCmd(VCS_COMMAND, "--version") } )
        if (gradleVersionNumber != null && gradleVersionNumber >= SYS_PROPERTIES_VERSION)  {
            osArch.set(providerFactory.systemProperty(MK_OS_ARCH) )
            osName.set(providerFactory.systemProperty(MK_OS_NAME) )
            osVersion.set(providerFactory.systemProperty(MK_OS_VERSION) )
            javaVersion.set(providerFactory.systemProperty(MK_JAVA_VERSION))
        } else {
            osArch.set(providerFactory.provider { System.getProperty(MK_OS_ARCH) } )
            osName.set(providerFactory.provider { System.getProperty(MK_OS_NAME) } )
            osVersion.set(providerFactory.provider { System.getProperty(MK_OS_VERSION) } )
            javaVersion.set(providerFactory.provider { System.getProperty(MK_JAVA_VERSION) })
        }
    }

    companion object {
        private val VALID_VCS_PROVIDERS: Collection<String> = listOf("github-enterprise",
            "bitbucket-server", "gitlab-onpremise", "bitbucket", "github", "gitlab")
        private const val MK_OS_ARCH = "os.arch"
        private const val MK_OS_NAME = "os.name"
        private const val MK_OS_VERSION = "os.version"
        private const val MK_JAVA_VERSION = "java.version"
        private const val VCS_COMMAND = "git"
        private const val CHARSET_UTF8 = "UTF-8"
        private val SYS_PROPERTIES_VERSION = VersionNumber.parse("6.1")

        @JvmStatic
        fun isValidVcsProvider(provider: String?): Boolean {
            return provider == null || VALID_VCS_PROVIDERS.contains(provider)
        }

        @JvmStatic
        fun parseProviderUrl(url: String?): String? {
            if (url != null) {
                for (provider: String in VALID_VCS_PROVIDERS) {
                    if (url.contains((provider))) {
                        return provider
                    }
                }
            }
            return null
        }

        private fun createService(
            timeoutDuration: Duration
        ): BugsnagReleasesService {
            return createService(OkHttpClient.Builder()
                .connectTimeout(timeoutDuration)
                .callTimeout(timeoutDuration)
                .build())
        }

        internal fun createService(
            okHttpClient: OkHttpClient
        ): BugsnagReleasesService {
            return Retrofit.Builder()
                .baseUrl("https://upload.bugsnag.com") // Not actually used
                .validateEagerly(true)
                .callFactory(okHttpClient)
                .addConverterFactory(ScalarsConverterFactory.create())
                .addConverterFactory(MoshiConverterFactory.create())
                .build()
                .create()
        }
    }
}

@JsonClass(generateAdapter = true)
internal data class ReleasePayload(
    val buildTool: String,
    val apiKey: String,
    val appVersion: String,
    val appVersionCode: String,
    val metadata: Map<String, String?>,
    val sourceControl: Map<String, String?>,
    val builderName: String?
)

internal interface BugsnagReleasesService {
    @POST
    fun upload(
        @Url endpoint: String,
        @Header("Bugsnag-Api-Key") apiKey: String,
        @Body payload: ReleasePayload
    ): retrofit2.Call<String>
}
